File: src/main/java/myLang/interpreter/Variable.java
package myLang.interpreter;

import myLang.response.MyLangException;

/**
 * Класс реализует простую абстракию для реализации переменных.
 * Содержит имя {@code name} и тип {@code T}.
 *
 * @param <T> тип переменной
 */
public class Variable<T> {
    private final int MAX_SYMBOL_COUNT = 40;
    private final String name;
    private T value;

    public Variable(String name, T value) throws MyLangException {
        if(name.length() > MAX_SYMBOL_COUNT){
            throw new MyLangException("Имя переменной " + name + " содержит больше " + MAX_SYMBOL_COUNT + " символов");
        }
        this.name = name;
        this.value = value;
    }

    public String getName() {
        return name;
    }

    public T getValue() {
        return value;
    }

    public void updateValue(T newValue){
        value = newValue;
    }

    @Override
    public boolean equals(Object obj) {
        if(obj instanceof Variable){
            Variable variable = (Variable) obj;
            return name.equals(variable.name);
        }
        return false;
    }

    @Override
    public int hashCode() {
        return name.hashCode();
    }

}

File: src/main/kotlin/myLangKotlin/interpreter/Variable.kt
package myLangKotlin.interpreter

import myLangKotlin.response.MyLangException

/**
 * Класс реализует простую абстракию для реализации переменных.
 * Содержит имя `name` и тип `T`.
 *
 * @param <T> тип переменной
 */
class Variable<T> @Throws(MyLangException::class)
constructor(val name: String, value: T) {
    private val MAX_SYMBOL_COUNT = 40
    var value: T? = null
        private set

    init {
        if (name.length > MAX_SYMBOL_COUNT) {
            throw MyLangException("Имя переменной $name содержит больше $MAX_SYMBOL_COUNT символов")
        }
        this.value = value
    }

    fun updateValue(newValue: T) {
        value = newValue
    }

    override fun equals(other: Any?): Boolean {
        if (other is Variable<*>) {
            val variable = other as Variable<*>?
            return name == variable!!.name
        }
        return false
    }

    override fun hashCode(): Int {
        return name.hashCode()
    }

}
